src/utils/dom-fingerprint.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+
+/**
+ * Gets a string of concatenated element names and relative positions
+ * of the DOM element and its parentElement's (up to 25).  Relative position
+ * is the index of nodes with this tag within the parent's children.
+ * The order is from the inner to outer nodes in DOM hierarchy.
+ *
+ * If a DOM hierarchy is the following:
+ *
+ * <div id='id1' ...>
+ *   <div id='id2' ...>
+ *     <table ...>       // table:0
+ *       <tr>            // tr:0
+ *         <td>...</td>  // td:0
+ *         <td>          // td:1
+ *           <amp-ad ...></amp-ad>
+ *         </td>
+ *       </tr>
+ *       <tr>...</tr>    // tr:1
+ *     </table>
+ *   </div>
+ * </div>
+ *
+ * With the amp-ad element passed in:
+ * 'amp-ad.0,td.1,tr.0,table.0,div/id2.0,div/id1.0'
```

Comment:
No particular reason. We always end up hashing the string, so as long as it's unique, it doesn't matter.